File: 0411-Minimum Unique Word Abbreviation/0411-Minimum Unique Word Abbreviation.cpp
class Solution {
public:
    string minAbbreviation(string target, vector<string>& dictionary) {
        int m = target.size();
        vector<int> dict;
        int candidate = 0;
        for (string s : dictionary) {
            if (s.size() != m) {
                continue;
            }
            int mask = 0;
            for (int i = m - 1, b = 1; i >= 0; --i, b <<= 1) {
                if (s[i] != target[i]) {
                    mask |= b;
                }
            }
            dict.push_back(mask);
            candidate |= mask;
        }
        
        int bm = 1 << m;
        int minLen = m;
        int minAbbr = (1 << m) - 1;
        dfs(dict, 1, 0, bm, candidate, 1, minLen, minAbbr);

        string result;
        int prev = 0;
        int b = (1 << (m - 1));
        for (int i = 0; i < m; ++i) {
            if (minAbbr & b) {
                if (i - prev > 0) {
                    result += to_string(i - prev);
                }
                prev = i + 1;
                result += target[i];
            }
            else if (i == m - 1) {
                result += to_string(m - prev);
            }
            b >>= 1;
        }
        
        return result;
    }
    
private:
    void dfs(vector<int>& dict, int start, int mask, int bm, int candidate, int len, int& minLen, int& minAbbr) {
        if (len >= minLen) {
            return;
        }
        
        bool match = true;
        for (int dm : dict) {
            if ((mask & dm) == 0) {
                match = false;
                break;
            }
        }
        
        if (match) {
            minLen = len;
            minAbbr = mask;
        }
        else {
            for (int b = start; b < bm; b <<= 1) {
                if (candidate & b) {
                    int nextLen;
                    if (b == start && (b << 1) == bm) {
                        nextLen = len;
                    }
                    else if (b == start || (b << 1) == bm) {
                        nextLen = len + 1;
                    }
                    else {
                        nextLen = len + 2;
                    }
                    dfs(dict, b << 1, mask | b, bm, candidate, nextLen, minLen, minAbbr);
                }
            }
        }
    }
};

File: 0401-0500/0411-Minimum Unique Word Abbreviation/0411-Minimum Unique Word Abbreviation.py
class Solution:
    def minAbbreviation(self, target: str, dictionary: List[str]) -> str:
        m = len(target)
        table = []
        candidate = 0
        for s in dictionary:
            if len(s) != m:
                continue
            mask = 0
            b = 1
            for i in range(m - 1, -1, -1):
                if s[i] != target[i]:
                    mask |= b
                b <<= 1
            table.append(mask)
            candidate |= mask
        
        bm = 1 << m
        minLen = m
        minAbbr = (1 << m) - 1
        def dfs(start, mask, l):
            nonlocal minLen, minAbbr
            if l >= minLen:
                return
            
            if any(mask & tm == 0 for tm in table):
                b = start
                while b < bm:
                    if (candidate & b):
                        if b == start and (b << 1) == bm:
                            nextLen = l
                        elif b == start or (b << 1) == bm:
                            nextLen = l + 1
                        else:
                            nextLen = l + 2
                        dfs(b << 1, mask | b, nextLen)
                    b <<= 1
            else:
                minLen = l
                minAbbr = mask

        dfs(1, 0, 1)
        result = []
        prev = 0
        b = 1 << (m - 1)
        for i in range(m):
            if minAbbr & b:
                if i - prev > 0:
                    result.append(str(i - prev))
                prev = i + 1
                result.append(target[i])
            elif i == m - 1:
                result.append(str(m - prev))
            b >>= 1

        return ''.join(result)
